### PR TITLE
Multi-J Algo: Make div(E)/div(B) Cleaning Optional

### PR DESCRIPTION
**Pull request by EZoni**
With this PR it is not necessary anymore to always set `warpx.do_dive_cleaning = 1` and `warpx.do_divb_cleaning = 1` when `psatd.J_linear_in_time = 1`. The F and G correction terms in the PSATD equations corresponding to the `psatd.J_linear_in_time` assumption are therefore optional and added only if needed.

### CODE REVIEW

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithm.cpp [hunk 1]
```diff
@@ -270,33 +288,46 @@ PsatdAlgorithm::pushSpectralFields (SpectralFieldData& f) const
...
+
+                fields(i,j,k,Idx.Ey) += -X1 * (Jy_new - Jy) / dt;
+
+                fields(i,j,k,Idx.Ez) += -X1 * (Jz_new - Jz) / dt;
+
+                fields(i,j,k,Idx.Bx) += I * X2/c2 * (ky * (Jz_new - Jz) - kz * (Jy_new - Jy));
+
+                fields(i,j,k,Idx.By) += I * X2/c2 * (kz * (Jx_new - Jx) - kx * (Jz_new - Jz));
+
+                fields(i,j,k,Idx.Bz) += I * X2/c2 * (kx * (Jy_new - Jy) - ky * (Jx_new - Jx));
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithm.cpp]: Just a bit of taste - at some point too many newlines can hinder readability again :)
```suggestion
                fields(i,j,k,Idx.Ex) += -X1 * (Jx_new - Jx) / dt;
                fields(i,j,k,Idx.Ey) += -X1 * (Jy_new - Jy) / dt;
                fields(i,j,k,Idx.Ez) += -X1 * (Jz_new - Jz) / dt;

                fields(i,j,k,Idx.Bx) += I * X2/c2 * (ky * (Jz_new - Jz) - kz * (Jy_new - Jy));
                fields(i,j,k,Idx.By) += I * X2/c2 * (kz * (Jx_new - Jx) - kx * (Jz_new - Jz));
                fields(i,j,k,Idx.Bz) += I * X2/c2 * (kx * (Jy_new - Jy) - ky * (Jx_new - Jx));
```

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithm.cpp [hunk 2]
```diff
@@ -309,33 +340,45 @@ PsatdAlgorithm::pushSpectralFields (SpectralFieldData& f) const
...
+                        fields(i,j,k,Idx.Ez_avg) += I * c2 * ep0 * X1 * F_old * kz;
+                    }
+
+                    if (divb_cleaning)
+                    {
+                        fields(i,j,k,Idx.Bx_avg) += I * c2 * ep0 * X1 * G_old * kx;
+
+                        fields(i,j,k,Idx.By_avg) += I * c2 * ep0 * X1 * G_old * ky;
+
+                        fields(i,j,k,Idx.Bz_avg) += I * c2 * ep0 * X1 * G_old * kz;
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithm.cpp]: Just curious about the spacious use of newlines :)
(same for the blocks of `dive_cleaning` and `divb_cleaning` above) - what do you think?
```suggestion
                        fields(i,j,k,Idx.Ex_avg) += I * c2 * ep0 * X1 * F_old * kx;
                        fields(i,j,k,Idx.Ey_avg) += I * c2 * ep0 * X1 * F_old * ky;
                        fields(i,j,k,Idx.Ez_avg) += I * c2 * ep0 * X1 * F_old * kz;
                    }

                    if (divb_cleaning)
                    {
                        fields(i,j,k,Idx.Bx_avg) += I * c2 * ep0 * X1 * G_old * kx;
                        fields(i,j,k,Idx.By_avg) += I * c2 * ep0 * X1 * G_old * ky;
                        fields(i,j,k,Idx.Bz_avg) += I * c2 * ep0 * X1 * G_old * kz;
```

File: Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp [hunk 3]
```diff
@@ -66,8 +66,15 @@ SpectralFieldIndex::SpectralFieldIndex (const bool update_with_rho,
         {
             Jx_new = c++; Jy_new = c++; Jz_new = c++;
 
-            // TODO Allocate F and G only when needed
-            F = c++; G = c++;
+            if (dive_cleaning)
+            {
+                F = c++;
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp:71]: Just double checking for all the values here: you want to assign `c` and then increase `c` afterwards, right?

Comment (EZoni) [reply Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp:71]: Yes, that's right. First assign, and then increase.

**Comment by EZoni** [comment]
@ax3l I removed the unnecessary newlines in https://github.com/ECP-WarpX/WarpX/pull/2116/commits/977adbf7bc81c077b9967c6a7471110e934a5c7e (I think there were newlines before the equations got split into smaller parts).

**Review by ax3l — APPROVED** [review]
Thank you! :tada:

**Pull request merged**